multiuser/api/che-multiuser-api-authentication-commons/src/main/java/org/eclipse/che/multiuser/api/authentication/commons/filter/MultiUserEnvironmentInitializationFilter.java

```diff
@@ -0,0 +1,173 @@
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.multiuser.api.authentication.commons.filter;
+
+import static org.eclipse.che.multiuser.api.authentication.commons.Constants.CHE_SUBJECT_ATTRIBUTE;
+
+import java.io.IOException;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import org.eclipse.che.commons.auth.token.RequestTokenExtractor;
+import org.eclipse.che.commons.env.EnvironmentContext;
+import org.eclipse.che.commons.subject.Subject;
+import org.eclipse.che.multiuser.api.authentication.commons.SessionStore;
+import org.eclipse.che.multiuser.api.authentication.commons.SubjectHttpRequestWrapper;
+
+/**
+ * Performs basic environment initialization actions as follows:
+ *
+ * <ul>
+ *   <li>Extracts token from request
+ *   <li>Checks token for validity and fetch user Id (implementation specific)
+ *   <li>Fetch cached {@link HttpSession} and {@link Subject} or construct new (implementation
+ *       specific))
+ *   <li>Set subject for current request into {@link EnvironmentContext}
+ * </ul>
+ *
+ * @author Max Shaposhnyk (mshaposh@redhat.com)
+ */
+public abstract class MultiUserEnvironmentInitializationFilter implements Filter {
+
+  private final SessionStore sessionStore;
+  private final RequestTokenExtractor tokenExtractor;
+
+  public MultiUserEnvironmentInitializationFilter(
+      SessionStore sessionStore, RequestTokenExtractor tokenExtractor) {
+    this.sessionStore = sessionStore;
+    this.tokenExtractor = tokenExtractor;
+  }
+
+  /**
+   * Wraps {@link HttpServletRequest} to handle HTTP session creation requests and return cached one
+   * if possible.
+   */
+  protected class SessionCachedHttpRequest extends HttpServletRequestWrapper {
+
+    /**
+     * Constructs a request object wrapping the given request.
+     *
+     * @throws IllegalArgumentException if the request is null
+     */
+    private final String userId;
+
+    public SessionCachedHttpRequest(ServletRequest request, String userId) {
+      super((HttpServletRequest) request);
+      this.userId = userId;
+    }
+
+    @Override
+    public HttpSession getSession() {
+      return getOrCreateSession(true);
+    }
+
+    @Override
+    public HttpSession getSession(boolean create) {
+      return getOrCreateSession(create);
+    }
+
+    /* Finds cached session or creates new if allowed */
+    private HttpSession getOrCreateSession(boolean createNew) {
```

Comment:
I see no value in this private method, I would inline it directly in `getSession`.
The behavior is already explained in `HttpServletRequest` interface, so you can skip the javadoc or document your implementation and reference the interface doc. But I think you're not doing anything special here.

Reply:
yes, interesting. Does it have any consequence? What do we store in the session?

Reply:
Inilining is not good here since there are two getSession() methods. But anyway i will add an extra explanation about sessions.

Reply:
Added not into class javadoc

Reply:
ah, I didn't noticed 2 calls. I would still prefer to inline it and `getSession()` would call `getSession(true)`. But it's personal preference so I'm ok with that.